geonotebook/layers.py

```diff
@@ -1,14 +1,98 @@
 from collections import namedtuple
 from collections import OrderedDict
 
+import sys
+
 import six
 
 from . import annotations
 from .config import Config
+from .vis.utils import generate_colormap
+
 
 BBox = namedtuple('BBox', ['ulx', 'uly', 'lrx', 'lry'])
 
 
+class VisOptions(object):
+    def __init__(self, data, opacity=1.0, gamma=1.0, projection='EPSG:3857',
+                 kernel_id=None, zIndex=None, colormap=None, interval=None,
+                 layer_type=None, **kwargs):
+
+        # self.vis_url = vis_url
+        self.opacity = opacity
+        self.projection = projection
+        self.gamma = gamma
+        self.interval = interval
+        self.colormap = None
+        self.zIndex = zIndex
+        self.kernel_id = kernel_id
+        self.layer_type = layer_type
+
+        try:
+            num_bands = len(data.band_indexes)
+        except AttributeError:
+            num_bands = -1
+
+        if num_bands == 1:
+            try:
+                # If we have the colormap in the form
+                # of a list of dicts with color/quantity then
+                # set options['colormap'] equal to this
+                for d in colormap:
+                    assert 'color' in d
+                    assert 'quantity' in d
+
+                self.colormap = colormap
+
+            except Exception:
+                # Otherwise try to figure out the correct colormap
+                # using data min/max
+                try:
+                    _min, _max = interval
+                except TypeError:
+                    _min, _max = None, None
+
+                if _min is None:
+                    try:
+                        _min = min(data.min)
+                    except (ValueError, TypeError):
+                        _min = data.min
+
+                if _max is None:
+                    try:
+                        _max = max(data.max)
+                    except (ValueError, TypeError):
+                        _max = data.max
+
+                self.colormap = generate_colormap(
+                    colormap, _min, _max)
+        else:
+            self.colormap = []
+
+    def serialize(self):
```

Comment:
```serialize``` is implemented on a bunch of different classes right now,  I'd like to keep their general signature/return type the same for consistency sake (and for later refactoring).   Currently they are returning dicts  because in the case of some of these classes we expect to call ```serialize()``` on values of an object being serialized  (e.g.  GeonotebookLayerCollection.serialize calls serialize on each of its GeonotebookLayers).